client/src/components/CosmologyForm.vue

```diff
@@ -13,90 +13,138 @@
+      required: true,
+    },
   },
-  data: () => ({
-    inputs: {
-      H0: {
-        html: '<label>H<sub>0</sub></label>',
-        min: 10,
-        max: 500,
-        step: 1,
-      },
-      Ob0: {
-        html: '<label>&#937;<sub>b</sub></label>',
-        min: 0.005,
-        max: 0.65,
-        step: 0.001,
-      },
-      Om0: {
-        html: '<label>&#937;<sub>m</sub></label>',
-        min: 0.02,
-        max: 2,
-        step: 0.01,
+  model: {
+    prop: 'cosmoData',
+    event: 'updateCosmo',
+  },
+  data() {
+    return {
+      inputs: {
+        H0: {
+          html: 'H<sub>0</sub>',
+          min: 10,
+          max: 500,
+          step: 1,
+        },
+        Ob0: {
+          html: '&#937;<sub>b</sub>',
+          min: 0.005,
+          max: 0.65,
+          step: 0.001,
+        },
+        Om0: {
+          html: '&#937;<sub>m</sub>',
+          min: 0.02,
+          max: 2,
+          step: 0.01,
+        },
+        z: {
+          html: 'Redshift',
+          min: 0,
+          max: 1100,
+        },
+        n: {
+          html: 'n<sub>s</sub>',
+          min: -4,
+          max: 3,
+          helpText: 'Spectral Index',
+        },
+        sigma_8: {
+          html: '&#963<sub>8</sub>',
+          min: 0.1,
+          helpText: 'RMS Mass Fluctuations',
+        },
       },
-    },
-    cosmologyChoices,
-    cosmologyChoice: cosmologyChoices[0],
-  }),
+      /**
+       * Represents the different selections of the cosmo model and saves the
+       * users' inputs for each.
+       */
+      allCosmoData: JSON.parse(JSON.stringify(BACKEND_CONSTANTS.cosmo_params)),
+      cosmologyChoices: Object.keys(BACKEND_CONSTANTS.cosmo_params),
+      cosmologyChoice: this.cosmoData.cosmo_model,
+    };
+  },
   components: {
-    FormNumberField,
+    DoubleField,
+  },
+  created() {
+    if (this.cosmoData.cosmo_params === null) {
+      const {
+        H0, Ob0, Om0, z, n, sigma_8,
+      } = BACKEND_CONSTANTS.cosmo_params.Planck13;
+      this.cosmoData.cosmo_params = {
+        H0,
+        Ob0,
+        Om0,
+      };
+      this.cosmoData.z = z;
+      this.cosmoData.n = n;
+      this.cosmoData.sigma_8 = sigma_8;
+      this.cosmoData.cosmo_model = 'Planck13';
+      this.cosmologyChoice = 'Planck13';
+      this.$emit('updateCosmo', this.cosmoData);
+    }
   },
   methods: {
     createSetCurrentValueFunc(inputType) {
       return (newValue) => {
-        // Set the new value temporarily
-        this.cosmoValues[inputType] = newValue;
+        if (this.cosmoData.cosmo_params[inputType]) {
+          this.cosmoData.cosmo_params[inputType] = newValue;
+        } else {
+          this.cosmoData[inputType] = newValue;
+        }
+
+        // Set the value in allCosmoData
+        this.allCosmoData[this.cosmoData.cosmo_model][inputType] = newValue;
 
         // Set the value for good.
-        this.setCosmo(this.cosmoValues);
+        this.$emit('updateCosmo', this.cosmoData);
       };
     },
```

Comment:
So the way this is set up currently, you're passing the cosmo object into your component by reference then setting the state of its values manually. v-model never actually does anything because by the time you hit your $emit call you've already set those data members in the parent.

One good solution to this is maintaining a local copy of params and sending a copy of that via the emit instead. This made me realize that my forms all need a minor tweak to avoid passing the actual local object up by reference, too.

Reply:
Setting a prop manually sets it in the parent? This can be changed to what is being suggested, but that doesn't seem right. https://stackoverflow.com/questions/43701175/is-it-possible-to-change-props-value-from-method-in-vue-component#:~:text=1%20Answer&text=What%20you%20are%20doing%20will,Vue%20(in%20the%20console).&text=The%20value%20will%20actually%20change,re%2Drenders%20for%20any%20reason.

Reply:
If it's a value it will pass a copy. If its an object / array it will pass by reference. Functions parameters work the same way in javascript!

https://vuejs.org/v2/guide/components-props.html
![image](https://user-images.githubusercontent.com/25331712/99194883-43da5a00-2737-11eb-9bbb-ffbbdc7d228a.png)